src/implementation/base.jl

```diff
@@ -182,44 +183,50 @@ function tensorcontract!(C::AbstractArray,
     soB1 = prod(soB)
     sc1 = prod(sc)
 
-    AB = tensoralloc_contract(eltype(C), A, pA, conjA, B, pB, conjB,
-                              trivialpermutation(pAB), true, allocator)
-    ÃB̃ = reshape(AB, (soA1, soB1))
+    AB′ = tensoralloc_contract(eltype(C), A, pA, conjA, B, pB, conjB,
+                               trivialpermutation(pAB), Val(true), allocator)
+    ÃB̃ = reshape(AB′, (soA1, soB1))
     if conjA && conjB
-        Atemp = tensoralloc_add(eltype(C), A, reverse(pA), conjA, true, allocator)
-        Btemp = tensoralloc_add(eltype(C), B, reverse(pB), conjB, true, allocator)
-        Ã = reshape(permutedims!(Atemp, A, linearize(reverse(pA))), (sc1, soA1))
-        B̃ = reshape(permutedims!(Btemp, B, linearize(reverse(pB))), (soB1, sc1))
-        mul!(ÃB̃, adjoint(Ã), adjoint(B̃))
+        Atemp′ = tensoralloc_add(eltype(C), A, reverse(pA), conjA, Val(true), allocator)
+        Btemp′ = tensoralloc_add(eltype(C), B, reverse(pB), conjB, Val(true), allocator)
+        Ãtemp = adjoint(reshape(permutedims!(Atemp′, A, linearize(reverse(pA))),
+                                 (sc1, soA1)))
+        B̃temp = adjoint(reshape(permutedims!(Btemp′, B, linearize(reverse(pB))),
+                                 (soB1, sc1)))
     elseif conjA
-        Atemp = tensoralloc_add(eltype(C), A, reverse(pA), conjA, true, allocator)
-        Btemp = tensoralloc_add(eltype(C), B, pB, conjB, true, allocator)
-        Ã = reshape(permutedims!(Atemp, A, linearize(reverse(pA))), (sc1, soA1))
-        B̃ = reshape(permutedims!(Btemp, B, linearize(pB)), (sc1, soB1))
-        mul!(ÃB̃, adjoint(Ã), B̃)
+        Atemp′ = tensoralloc_add(eltype(C), A, reverse(pA), conjA, Val(true), allocator)
+        Btemp′ = tensoralloc_add(eltype(C), B, pB, conjB, Val(true), allocator)
+        Ãtemp = adjoint(reshape(permutedims!(Atemp′, A, linearize(reverse(pA))),
+                                 (sc1, soA1)))
+        B̃temp = reshape(permutedims!(Btemp′, B, linearize(pB)), (sc1, soB1))
     elseif conjB
-        Atemp = tensoralloc_add(eltype(C), A, pA, conjA, true, allocator)
-        Btemp = tensoralloc_add(eltype(C), B, reverse(pB), conjB, true, allocator)
-        Ã = reshape(permutedims!(Atemp, A, linearize(pA)), (soA1, sc1))
-        B̃ = reshape(permutedims!(Btemp, B, linearize(reverse(pB))), (soB1, sc1))
+        Atemp′ = tensoralloc_add(eltype(C), A, pA, conjA, Val(true), allocator)
+        Btemp′ = tensoralloc_add(eltype(C), B, reverse(pB), conjB, Val(true), allocator)
+        Ã = reshape(permutedims!(Atemp′, A, linearize(pA)), (soA1, sc1))
+        B̃ = adjoint(reshape(permutedims!(Btemp′, B, linearize(reverse(pB))), (soB1, sc1)))
         mul!(ÃB̃, Ã, adjoint(B̃))
```

Comment:
Was there a reason for this restructuring? Also, there is a `mul!` still here above this comment that you forgot to remove, but I can do it. However, I kinda liked the old version where I see the adjoint in the `mul!` call?

Reply:
I think I mostly wanted to make sure that
- I call free on the exact object that is allocated 
- I avoid reassigning objects of different types to the same variable name
And then noticed that there's a lot of duplicate code. I don't have any strong opinions about this though, I don't think it matters too much

Reply:
Ok I see how adding `tensorfree!` either leads to a lot of code duplication if it is repeated within each of the cases, or alternatively, feels inconsistent if these calls are after the `if` statement, but the `mul!` call is within the different cases.